README.md
# apTreeshape

CHANGES
Changes in apTreeshape 1.4.5
============================

  o Add simulation and inference function for the ranked tree shape model by Malie, Gascuel & Lambert



Changes in apTreeshape 1.4.4
============================

bug fixes
---------
  o Add .Rd file for data: carnivora.treeshape cytochromc primates rhodopsini
  o Suppress COPYING file 
 


Changes in apTreeshape 1.4.3
============================

bug fixes
---------
  o Minor bug corrections in function maxlik.betasplit.



Changes in apTreeshape 1.4.2
============================

bug fixes
---------
  o Function dbtrees has been removed.

Changes in apTreeshape 1.4.0
============================

improvements
------------
  o New function added: maxlik.betasplit

bug fixes
---------
  o Minor bugs have been fixed. Improved compatibility with ape

Changes in apTreeshape 1.1.0
============================

improvements
------------

  o  A new model can be simulated (Aldous' Branching model)

  o  'plot.treeshape' has been changed to be more intuitive


changes in behaviour
--------------------

  o  Only one tree generator has been kept ('rtreeshape'). The others have been declared internal.
  
  o  Only one fonction to connect to Internet databases has been kept, and its name is now 'dbtrees'

  o  'dbtrees' defaults have been modified for argument 'class'

  o  printed messages have been modified for the function 'dbtrees'


bug fixes
--------

  o  compatibility between classes 'numeric' and 'integer' has been improved


R/aldous.test.R
#comment
"aldous.test" <-
function (tree, xmin = 20, ...)
{
    if (identical(tree, NULL)) {
        stop("invalid tree", "\n")
    }
    if (xmin < 0.1) {
        xmin = 0.1
    }
    clade = smaller.clade.spectrum(tree)
    if (xmin > clade[1, 1]) {
        stop("'xmin' greater than the size of the tree")
    }
    ##clade = clade[clade[, 1] > xmin, ]
    xlab = "Size of parent clade (log scale)"
    ylab = "Size of smaller daughter clade (log scale)"
    plot(clade, pch = 20, xlab = xlab, log = "xy", ylab = ylab,
        xlim = c(xmin, max(clade[1, ]) + 10), ...)
    if (xmin >= 10) {
        x = xmin + 5
        text(x = x, y = 3/2 + 0.1, label = "PDA model", col = 4)
        text(x = x, y = x/4 - 1, label = "Yule model", col = 4)
    }
    else {
        x = 20
        text(x = x, y = 3/2 + 0.1, label = "PDA model", col = 4)
        text(x = x, y = x/4 + 2, label = "Yule model", col = 4)
    }
    mod = rq(clade[, 2] ~ clade[, 1])
    x = sort(clade[, 1])
    coef = coefficients(mod)
    abline(coef,untf=TRUE, col = 3)
    abline(0, 1/4, untf = TRUE)
    abline(3/2, 0, untf = TRUE)
    abline(v = 30, lty = 3)
}



R/get_PD_sample.R
get_extinction_list=function(N,tree,equal.ab=TRUE){
  
  tip.ab=tree$tip.ab
  names(tip.ab) = tree$tip.label
  tips.order = NULL
  if (equal.ab == TRUE) {
    tips.order = sample(names(tip.ab), N)
    
  } else {  
    ab_null = c(which(is.na(tip.ab)), which(tip.ab==0))
    if (length(ab_null)>0) {
      ab_null_order = sample(ab_null, length(ab_null))
      tip_not_null = tip.ab[-ab_null]
      tips.order = c(names(ab_null_order), names(tip_not_null)[order(tip_not_null,names(tip_not_null))]) 
    } else {
      tips.order = names(tip.ab)[order(tip.ab,names(tip.ab))] 
    }
  } 
  return(tips.order)
}

get_PD_sample=function(epsilon,beta,alpha,N,sampl.frac,ntree,equal.ab,eta,lengths="yule",b=1,d=0){
  lambda=lambda_N(epsilon,beta,N,upper=10000)
  PD=pbsapply(1:ntree,function(i){
    if(lengths=="yule"){
      tree=simulate_yule(epsilon,alpha,beta,N,b = b,d=d,equal.ab = equal.ab,eta = eta,lambda=lambda)
    }else if(lengths=="kingman"){
      tree=simulate_kingman(epsilon,alpha,beta,N,equal.ab = equal.ab,eta = eta,lambda=lambda)
    }
    M=max(depth(tree))
    tip=get_extinction_list(N,tree,equal.ab)
    PD.ech=sapply(1:length(sampl.frac),function(j){
      if (round(sampl.frac[j]*N)==N){sum(depth(tree))+M
      } else if (round(sampl.frac[j]*N) == 0) {
        0
      } else if (round(sampl.frac[j]*N) == 1) {
        M
      } else {
        nd=depth(drop.tip(tree,tip[1:(N-round(sampl.frac[j]*N))]))
        sum(nd)+M}})
    return(PD.ech/(sum(depth(tree))+M))
  })
  rownames(PD) = sampl.frac
  colnames(PD) = seq(1,ntree)
  return(PD)
}


R/get_tree_beta.R
get_tree_beta=function(epsilon,beta,alpha,N,sampl.frac,ntree,equal.ab=TRUE,eta=1){
  beta=pbsapply(1:ntree,function(i){
    tree=simulate_tree(epsilon, alpha, beta, N, equal.ab=equal.ab, eta=eta)
    tree$tip.label=1:length(tree$tip.label)
    tip=get_extinction_list(N,tree,equal.ab)
    return(sapply(1:length(sampl.frac),function(j){
      if (round(sampl.frac[j]*N)==N){maxlik.betasplit(as.treeshape(tree), up=10, remove.outgroup=FALSE, confidence.interval="none")$max_lik
      } else if (round(sampl.frac[j]*N)==0 || round(sampl.frac[j]*N)==1) {
        NA
      } else {
        maxlik.betasplit(as.treeshape(drop.tip(tree,tip[1:(N-round(sampl.frac[j]*N))])), up=10, remove.outgroup=FALSE, confidence.interval="none")$max_lik       
      }
    }))
  })
  rownames(beta) = sampl.frac
  colnames(beta) = seq(1,ntree)
  return(beta)
}


R/simulate_tree.R
lambda.epsilon=function(epsilon,beta,n,upper=10000){
  f=function(x){
    sapply(x,function(e){exp(-(beta+n+1)*e)*(1-exp(-e))^beta})
  }
  # integrate(f,epsilon,upper)
  int=2*adaptIntegrate(f, lowerLimit = epsilon, upperLimit = upper)$integral
  while(upper>0.1 & int==0){
    upper=upper/1.5
    int=2*adaptIntegrate(f, lowerLimit = epsilon, upperLimit = upper)$integral
  }
  return(int)
}

simulate.Yi=function(N,epsilon,beta,n){
  res=rep(0,N)
  if(N<1) return(c())
  
  if(beta<=0){
    g=function(x){((1-exp(-epsilon))^beta)*exp(-(beta+n+1)*epsilon)}  
    
    f=function(x){
      sapply(x,function(e){(1-exp(-e))^beta})
    }
    
    fun=function(i){
      continue=TRUE
      while(continue){
        e=rexp(1,(beta+n+1))+epsilon
        u=runif(1,0,1)
        if(u<(f(e)/g(e))){
          return(e)
          continue=FALSE
        }
      }
    }
    
    res=sapply(1:N,fun)
    
  }else{
    
    simulate.g=function(N,beta,n){
      res=rep(0,N)
      A=(beta+n+1)
      M=((beta/(A+beta))^beta)*(1+beta/A)^(-A)
      for(i in 1:N){
        u=runif(1,0,1)
        if(u<(-log(M)/(1-log(M)))){
          res[i]=runif(1,0,-log(M)/A)
        }else{
          res[i]=-log(M)/A+rexp(1,A)
        }
      }
      return(res)
    }
    
    f=function(x){
      A=beta+n+1
      M=((beta/(A+beta))^beta)*(1+beta/A)^(-A)
      sapply(x,function(e){exp(-A*e)*(1-exp(-e))^beta})
    }
    
    g=function(x){
      A=beta+n+1
      M=((beta/(A+beta))^beta)*(1+beta/A)^(-A)
      sapply(x,function(y){
        min(exp(-A*y),M)
      })
    }
    
    fun=function(i){
      continue=TRUE
      while(continue){
        e=simulate.g(1,beta,n)
        if(e>epsilon){
          u=runif(1,0,1)
          if(u<(f(e)/g(e))){
            return(e)
            continue=FALSE
          }}
      }
    }
    res=sapply(1:N,fun)
  }
  
  return(res)
}

simulate.Tau.X=function(epsilon,x,alpha,beta,n,ab=FALSE,eta=1,x.ab=1,lambda=NULL){
  if(is.null(lambda)){
    lambda=lambda.epsilon(epsilon,beta,n)
  }else{
    lambda=lambda[n]
  }
  an=sum(sapply(1:(n-1),function(i){beta(beta+1+i,beta+1+n-i)/((n+1)*beta(1+i,1+n-i))}))
  sigma=rexp(1,an)
  N=rpois(1,lambda*sigma)
  Y=c(simulate.Yi(N,epsilon,beta,n))
  Z=cumsum(c(0,Y))
  si=diff(c(0,sort(runif(N,0,sigma)),sigma))
  Tau=x^(-alpha)*(sum(exp(alpha*Z)*si))
  X=x*exp(-Z[length(Z)])
  if (ab){
    if(N==0){
      X.ab=x.ab
      return(c(Tau,X,X.ab))
    }
    
    if(eta==0){
      X.ab=x.ab/(2^N)
    }else{
      Y2=sum(log(exp(-Y*eta)+(1-exp(-Y))^eta))
      X.ab=x.ab*exp(-Z[length(Z)]*eta)/exp(Y2)
    }
    return(c(Tau,X,X.ab))
  }
  return(c(Tau,X))
}

simulate.R.K=function(beta,n){
  prob=sapply(1:(n-1),function(i){beta(beta+1+i,beta+1+n-i)/beta(1+i,1+n-i)})
  k=sample(1:(n-1),size=1,prob=prob)
  r=rbeta(1,beta+1+k,beta+1+n-k)
  return(c(r,k))
}

simulate.R=function(beta,n,K){
  r=rbeta(1,beta+1+K,beta+1+n-K)
  return(c(r,K))
}

bind.trees=function(a1,a2,root1,root2,ab=FALSE){
  if(a1$Nnode==0){return(a2)
  }else if (a2$Nnode==0){return(a1)
  }else{
    a1$root.edge=root1
    a2$root.edge=root2
    a=a1+a2
    if(ab){a$tip.ab=c(a1$tip.ab,a2$tip.ab)}
    return(a)
  }
}

simulate_tree=function(epsilon,alpha,beta,N,equal.ab=TRUE,eta=1,lambda=NULL){
  aux=function(x,n,x.ab=1){
    if(n>1){
      v1=simulate.R.K(beta,n)
      n1=v1[2]
      n2=n-n1
      x1=v1[1]*x
      x2=x-x1
      if(n1>1){
        if(equal.ab){v2.1=simulate.Tau.X(epsilon,x1,alpha,beta,n1,lambda=lambda)
        }else{v2.1=simulate.Tau.X(epsilon,x1,alpha,beta,n1,ab=TRUE,eta = eta,x.ab=(x1^eta)/(x1^eta+x2^eta),lambda=lambda)}
        tree1=aux(v2.1[2],n1)
        if(!equal.ab){
          tree1$tip.ab=v2.1[3]*tree1$tip.ab
        }
        root1=v2.1[1]
      }else{
        tree1=list(edge=matrix(c(2,1),1,2),edge.length=0,Nnode=1,tip.label=NA,tip.ab=c(1))
        if(!equal.ab){
          tree1$tip.ab=((x1^eta)/(x1^eta+x2^eta))
        }
        class(tree1)="phylo"
        
        root1=1
      }
      if(n2>1){
        if(equal.ab){v2.2=simulate.Tau.X(epsilon,x2,alpha,beta,n2,lambda=lambda)
        }else{v2.2=simulate.Tau.X(epsilon,x2,alpha,beta,n2,ab=TRUE,eta = eta,x.ab=(x2^eta)/(x1^eta+x2^eta),lambda=lambda)}
        tree2=aux(v2.2[2],n2)
        if(!equal.ab){
          tree2$tip.ab=v2.2[3]*tree2$tip.ab
        }
        root2=v2.2[1]
      }else{
        tree2=list(edge=matrix(c(2,1),1,2),edge.length=0,Nnode=1,tip.label=NA,tip.ab=c(1))
        if(!equal.ab){
          tree2$tip.ab=((x2^eta)/(x1^eta+x2^eta))
        }
        class(tree2)="phylo"
        
        root2=1
      }
      a=bind.trees(tree1,tree2,root1,root2,ab=TRUE)
      A=collapse.singles(a)
      A$tip.ab=a$tip.ab
      return(A)
    }else{
      tree=list(edge=matrix(c(2,1),1,2),edge.length=1,Nnode=1,tip.label=NA,tip.ab=1)
      class(tree)="phylo"
      return(tree)
    }
  }
  
  tree=aux(1,N)
  order=c(rep(N,N),rank(node.depth.edgelength(tree)[(N+1):(2*N-1)]))
  for(i in 1:(2*N-2)){
    tree$edge.length[i]=order[tree$edge[i,2]]-order[tree$edge[i,1]]
  }
  return(tree)
}


R/simulate_kingman.R
build_tree=function(H,tip.lab=rep(NA,length(H)+1)){
  if (length(H)==0) {
    tree=list(edge=matrix(c(2,1),1,2),edge.length=0,Nnode=1,tip.label=tip.lab)
    class(tree)="phylo"
    return(tree)
  }else{
    H.crown=max(H)
    split=which.max(H)
    t1=tip.lab[1:split]
    t2=tip.lab[(split+1):length(tip.lab)]
    if (split>1) {H1=H[1:(split-1)]
    }else{H1=c()}
    if (split<length(H)){H2=H[(split+1):length(H)]
    }else {H2=c()}
    
    a1=build_tree(H1,t1)
    a2=build_tree(H2,t2)
    root1=H.crown-max(c(H1,0))
    root2=H.crown-max(c(H2,0))
    return(collapse.singles(bind.trees(a1,a2,root1,root2)))
  }
}

depth=function(tree){
  d=node.depth.edgelength(tree)
  d[1]-d[(tree$Nnode+2):(2*tree$Nnode+1)]
}

nodes_depths_ordonnes=function(tree){
  aux=function(phylo,noeud=phylo$Nnode+2){
    if(phylo$Nnode<=1){return(max(phylo$edge.length))
    }else{
      fils=phylo$edge[phylo$edge[,1]==noeud,][,2]
      if (fils[1]<(phylo$Nnode+2)){
        ad=extract.clade(phylo,fils[2])
        return(cbind(max(phylo$edge.length[phylo$edge[,1]==noeud]),aux(ad)))
      }else if (fils[2]<(phylo$Nnode+2)){
        ag=extract.clade(phylo,fils[1])
        return(cbind(aux(ag),max(phylo$edge.length[phylo$edge[,1]==noeud])))
      }else{
        ag=extract.clade(phylo,fils[1])
        ad=extract.clade(phylo,fils[2])
        hg=aux(ag)
        hd=aux(ad)
        h=max(hg)+phylo$edge.length[phylo$edge[,2]==fils[1]]
        return(cbind(hg,h,hd))
      }
    }
  }
  H=aux(tree)
  return(H)
}

simulate_kingman=function(epsilon,alpha,beta,N,equal.ab=TRUE,eta=1,lambda=NULL){
  tree=simulate_tree(epsilon,alpha,beta,N,equal.ab,eta,lambda=lambda)
  ab=tree$tip.ab
  order=rank(nodes_depths_ordonnes(tree))
  depths=cumsum(rexp(N-1,sapply((N:2),function(i){i*(i-1)/2})))
  tree=build_tree(depths[order],1:N)
  tree$tip.ab=ab
  return(tree)
}


R/mcmc_alpha.R
lambda_N=function(epsilon,beta,N,upper=10000){c(0,sapply(2:N, function(x){lambda.epsilon(epsilon,beta,x,upper)}))}

a_N=function(beta,N){c(0,sapply(2:N,function(n){sum(sapply(1:(n-1),function(j){beta(beta+1+j,beta+1+n-j)/((n+1)*beta(1+j,1+n-j))}))}))}

aux_lik=function(M,beta,alpha){
  n=nrow(M)
  int=M[,"int"]
  depths=M[,"depth"]
  proba=0
  current=c(1)
  
  while (length(current)>0){
    i=which.max(depths[current])
    p=alpha*(int[current[i]])-log(sum(exp(int[current]*alpha)))
    proba=p+proba
    
    
    current=insert(current,
                   (1:n)[M[,"father"]==M[current[i],"node"]],
                   i,TRUE)
  }
  return(proba)
}

split=function(tree){
  n=tree$Nnode+1
  rep=matrix(0,3,2*n-1)
  rep[1,1:n]=rep(1,n)
  for (i in (2*n-2):1){
    rep[1,tree$edge[i,1]]=rep[1,tree$edge[i,1]]+rep[1,tree$edge[i,2]]
    if (rep[3,tree$edge[i,1]]==0){
      rep[3,tree$edge[i,1]]=rep[1,tree$edge[i,2]]
    }else{
      rep[2,tree$edge[i,1]]=rep[1,tree$edge[i,2]]
    }
  }
  return(rep)
}

insert=function(v,e,i,replace){
  if (i==0){
    c(e,v)
  }else  if(i==1){
    if (replace){
      c(e,v[-1])
    }else{c(v[1],e,v[-1])}
  }else if (i==length(v)){
    if (replace){
      c(v[-length(v)],e)
    }else{
      c(v,e)
    }
  }else if (replace){
    c(v[1:(i-1)],e,v[(i+1):length(v)])
  }else{
    c(v[1:i],e,v[(i+1):length(v)])
  }
}

rbactrian <- function(n, m=0.95){
  mBactrian = m
  sBactrian = sqrt(1-m^2)
  z = mBactrian + rnorm(n,0,1)*sBactrian
  rdunif <- runif(n,0,1)<0.5
  sign <- ifelse(rdunif,-1,1)
  z=z*sign
  return(z)
}

transform=function(tree,unsampled,list_depth=NULL,change_depth=0){
  split=split(tree)
  ntip=tree$Nnode+1
  depth=c(rep(0,ntip),depth(tree))
  p_unsampled=0
  compute_depth=is.null(list_depth)
  if(compute_depth){list_depth=list(c())}
  fils=tree$edge[tree$edge[,1]==(ntip+1),2]
  M=c(ntip+1,-1,depth[ntip+1],ntip)
  k=2*ntip
  id.unsampled=1
  for (i in (ntip+2):(2*ntip-1)){
    father=tree$edge[tree$edge[,2]==i,1]
    node_father=father
    pf=depth[i]
    pp=depth[father]
    fils=tree$edge[tree$edge[,1]==i,2]
    if (unsampled[i]>0){
      if(compute_depth | (change_depth==i)){
        depths=runif(unsampled[i],pf,pp)
        depths=depths[order(depths,decreasing = TRUE)]
        list_depth[[i]]=depths
      }else{
        depths=list_depth[[i]]
      }
      for (j in 1:unsampled[i]){
        M=rbind(M,c(k,father,depths[j],split[1,i]))
        father=k
        k=k+1
        id.unsampled=id.unsampled+1
      }
      p_unsampled=p_unsampled+sum(log(1:unsampled[i]))-log(pp-pf)*unsampled[i]
    }
    M=rbind(M,c(i,father,pf,split[1,i]))
  }
  colnames(M)=c("node","father","depth","ntip")
  for(j in 1:M[1,"depth"]){
    N=sum(M[,"depth"]>(j-1) & M[,"depth"]<(j))
    if(N>0){
      p_unsampled=p_unsampled-sum(log(1:N))
    }
  }
  
  return(list(M=M,p_unsampled=p_unsampled,list_depth=list_depth))
}

enhance=function(tree,alpha,beta,epsilon,lambdaN,aN){
  N=tree$Nnode
  unsampled=c(-1,rep(-1,2*N))
  p_nUnsampled=0
  
  for (i in (N+3):(2*N+1)){
    n=extract.clade(tree,i)$Nnode+1
    
    if(n==1){unsampled[i]=-1
    }else {
      lambda=lambdaN[n]
      an=aN[n]
      sigma=rexp(1,an)
      unsampled[i]=rpois(1,lambda*sigma)
      p_nUnsampled=p_nUnsampled+log(an)-(unsampled[i]+1)*log(lambda+an)+(unsampled[i])*log(lambda)
      if(is.na(unsampled[i])) print(paste("a",alpha,"b",beta,"t",unsampled[i],"lamb",lambda,"an",an,"sig",sigma,"n",n))
    }
  }
  
  tr=transform(tree,unsampled)
  p_unsampled=tr$p_unsampled
  list_depth=tr$list_depth
  tr=tr$M
  
  int=c(0,rep(-Inf,nrow(tr)-1))
  Rs=c(1,rep(-1,nrow(tr)-1))
  
  for(i in 2:nrow(tr)){
    if(int[i]==-Inf){
      father=tr[i,2]
      j=(1:nrow(tr))[tr[,1]==father]
      if(father<(2*N+2)){
        R=simulate.R(beta,tr[j,4],tr[i,4])[1]
        int[i]=int[j]+log(R)
        Rs[i]=R
        if(i<nrow(tr)){
          k=((i+1):nrow(tr))[tr[(i+1):nrow(tr),2]==father]
          int[k]=int[j]+log(1-R)
          Rs[k]=(1-R)
        }
      }else{
        R=exp(-1*simulate.Yi(1,epsilon,beta,tr[i,4]))
        int[i]=int[j]+log(R)
        Rs[i]=R
      }
    }
  }
  
  
  tr=cbind(tr,int)
  return(list(transform=tr,unsampled=unsampled,int=int,Rs=Rs,p_nUnsampled=p_nUnsampled,p_unsampled=p_unsampled,list_depth=list_depth,Infinite=FALSE))
}

change_int=function(enhanced,ind,tree,alpha,beta,epsilon,lambdaN,aN){
  changeMax=5
  N=tree$Nnode
  unsampled=enhanced$unsampled
  int=enhanced$int
  p_nUnsampled=enhanced$p_nUnsampled
  list_depth=enhanced$list_depth
  continue=TRUE
  if(is.null(ind) ){
    rm=c()
  }else if (ind==(N+2)) {
    rm=c()
  }else{
    n=extract.clade(tree,ind)$Nnode+1
    rm=c()
    tr=enhanced$transform
    node=which(tr[,1]==ind)
    if(unsampled[ind]>0){
      rm=(node-1):(node-unsampled[ind])
      node=node-unsampled[ind]
    }
    
    if(n==1){unsampled[ind]=-1
    }else if(ind==(N+2)){unsampled[ind]=-1}else{
      lambda=lambdaN[n]
      an=aN[n]
      p_nUnsampled=p_nUnsampled+(unsampled[ind]+1)*log(lambda+an)-(unsampled[ind])*log(lambda)
      u=runif(1)
      if(u<(0.4)){
        unsampled[ind]=unsampled[ind]-1
      }else if(u>0.6){
        unsampled[ind]=unsampled[ind]+1
      }
      unsampled[ind]=unsampled[ind]+floor(rnorm(1,mean = 0.5,sd = 10))
      if(unsampled[ind]<0){
        proba.unsampled=-Inf
        continue=FALSE
      }else{
        p_nUnsampled=p_nUnsampled-(unsampled[ind]+1)*log(lambda+an)+(unsampled[ind])*log(lambda)
        continue=TRUE
      }
      if(is.na(unsampled[ind])) print(paste("a",alpha,"b",beta,"t",unsampled[ind],"lamb",lambda,"an",an,"sig",sigma,"n",n))
    }
  }
  
  if(! continue){
    enhanced$p_nUnsampled=-Inf
    return(enhanced)
  }
  
  tr=transform(tree,unsampled,list_depth = list_depth,change_depth = max(ind,0))
  p_unsampled=tr$p_unsampled
  list_depth=tr$list_depth
  tr=tr$M
  oldRs=c()
  
  if(length(rm)>0){
    oldRs=enhanced$Rs[rm]
    Rs=enhanced$Rs[-rm]
    depth=enhanced$transform[-rm,"depth"]
    
  }else{
    
    Rs=enhanced$Rs
    depth=enhanced$transform[,"depth"]
    
  }
  if(is.null(ind)){
    Rs=enhanced$Rs
  }else{if(unsampled[ind]>-1){
    add=c(-1,sample(c(oldRs,rep(-1,changeMax+max(0,unsampled[ind]-length(oldRs)))),unsampled[ind]))
    if(node<length(Rs)){
      Rs=c(Rs[1:(node-1)],add,Rs[(node+1):length(Rs)])
      depth=c(depth[1:(node-1)],tr[node:(node+unsampled[ind]),"depth"],depth[(node+1):length(depth)])
      
    }else{
      Rs=c(Rs[1:(node-1)],add)
      depth=c(depth[1:(node-1)],tr[node:(node+unsampled[ind]),"depth"])}
  }}
  
  tr[,"depth"]=depth
  
  int=c(0,rep(-Inf,nrow(tr)-1))
  if(! is.null(ind)){if(ind>(N+2)){father=tr[node,2]
  Ks=which(tr[,2]==father)
  j=(1:nrow(tr))[tr[,1]==father]
  if(Rs[Ks[1]]==1){R=simulate.R(beta,tr[j,4],tr[Ks[1],4])[1]
  Rs[Ks[1]]=R
  if(length(Ks)==2){Rs[Ks[2]]=(1-R)}}}}
  Rs[1]=1
  
  for(i in 2:nrow(tr)){
    if(int[i]==-Inf){
      father=tr[i,2]
      j=(1:nrow(tr))[tr[,1]==father]
      if(father<(2*N+2)){
        if(Rs[i]==-1){
          R=simulate.R(beta,tr[j,4],tr[i,4])[1]
          Rs[i]=R}
        R=Rs[i]
        int[i]=int[j]+log(R)
        if(i<nrow(tr)){
          k=((i+1):nrow(tr))[tr[(i+1):nrow(tr),2]==father]
          int[k]=int[j]+log(1-R)
          Rs[k]=(1-R)
        }
      }else{
        if(Rs[i]==-1){
          R=exp(-1*simulate.Yi(1,epsilon,beta,tr[i,4]))
          Rs[i]=R}
        R=Rs[i]
        int[i]=int[j]+log(R)
        
      }
    }
  }
  if(any(Rs<0) ) {print(ind); debug(change_int); change_int(enhance,ind,tree,alpha,beta,epsilon)}
  tr=cbind(tr,int)
  return(list(transform=tr,unsampled=unsampled,int=int,Rs=Rs,p_nUnsampled=p_nUnsampled,p_unsampled=p_unsampled,list_depth=list_depth,Infinite=FALSE))
}

mcmc_alpha=function(tree,epsilon,beta,niter,ini=0,V=0.1,chain=NULL,
                    verbose=10,silent=TRUE,Nadapt=Inf,NadaptMin=10,NadaptMax=Inf,
                    ma=-4,Ma=4,proposal="bactrian",accOpt=0.3,Vmin=0.001){
  
  
  ND=rank(nodes_depths_ordonnes(tree))
  tree2=build_tree(ND)
  tree$edge.length=tree2$edge.length
  N=tree$Nnode
  lambdaN=lambda_N(epsilon,beta,N+1)
  aN=a_N(beta,N+1)
  ntip=tree$Nnode+1
  ordre=rank(max(depth(tree))+1-depth(tree))
  X=rep(1,ntip-1)
  for (i in 1:(ntip-1)){X[ordre[i]]=((ntip+1):(ntip*2-1))[i]}
  
  
  posterior2=function(alpha,enhance){
    if(enhance$Infinite){return(-Inf)}
    Y=try(aux_lik(enhance$transform,beta,alpha))
    if(inherits(Y,"try-error")){Y=-Inf}
    p2=enhance$p_nUnsampled
    p3=enhance$p_unsampled
    if(any(is.na(c(Y,p2,p3))) | any(is.nan(c(Y,p2,p3))) | any(is.infinite(c(Y,p2,p3)))){return(-Inf)}
    return(Y-p3+p2)
  }
  
  posterior=function(alpha,enhance){
    if(enhance$Infinite ){return(-Inf)}
    Y=try(aux_lik(enhance$transform,beta,alpha))
    if(inherits(Y,"try-error")){Y=-Inf}
    if(is.na(Y) | is.nan(Y) | is.infinite(Y)){return(-Inf)}
    return(Y)
  }
  
  if (is.null(chain)){
    a=ini
    j=1
    post_actuel=-Inf
    while(post_actuel==-Inf){
      int=enhance(tree,a,beta,epsilon,lambdaN,aN)
      post_actuel=posterior(a,int)
      post_actuel_int=posterior2(a,int)
    }
    Mcmc=c(a,sum(int$unsampled[-(1:(N+2))]),post_actuel)
  }else{
    a=chain$a
    int=chain$int
    Mcmc=chain$mcmc
    j=nrow(chain$mcmc)
    V=chain$V
    post_actuel=chain$post_actuel
    post_actuel_int=posterior2(a,int)
  }
  n=0
  for(i in 1:niter){
    if((i+j) %% Nadapt ==0 & (i+j)>NadaptMin & (i+j)<NadaptMax){ 
      if(proposal=="uniform" | proposal=="bactrian"){
        acc=sum(diff(Mcmc[-c(1:floor(nrow(Mcmc)-Nadapt)),1])>0)/(Nadapt-1)
        if(acc<0.001){
          V=V/100
        }else if(acc>0.999){
          V=V*100
        }else{
          V = V * (tan((pi/2)*acc)/tan((pi/2)*accOpt))
        }
        V=max(V,Vmin)
      }else{
        V=max(Vmin,sqrt(var(Mcmc[-c(1:floor(nrow(Mcmc)/2)),1]))*2.36)
      }
      if(is.na(V)) V=1
    }
    
    #On tire les paramètres proposés : 
    post_actuel=try(posterior(a,int))
    if(proposal=="uniform"){
      new_a = a + (runif(1,0,1)-0.5)*3.4641016*V
    }else if (proposal=="bactrian"){
      new_a =a + rbactrian(1)*V
    }else{
      new_a=rnorm(1,a,V)}
    if(new_a>Ma| new_a<ma){
      post=-Inf
    }else{
      post=try(posterior(new_a,int))}
    
    if(post>=post_actuel){
      a=new_a
      post_actuel=post
      post_actuel_int=posterior2(a,int)
    }else{
      u=runif(1,0,1)
      if (log(u)<(post-post_actuel)){
        a=new_a
        post_actuel=post
        post_actuel_int=posterior2(a,int)
      }
    }
    
    ks=(N+3):(2*N+1)
    for(k in ks){
      new_int=change_int(int,k,tree,a,beta,epsilon,lambdaN,aN)
      post=posterior2(a,new_int)
      
      if(post>=post_actuel_int){
        int=new_int
        post_actuel_int=post
        post_actuel=posterior(a,int)
      }else{
        u=runif(1,0,1)
        if (log(u)<(post-post_actuel_int)){
          int=new_int
          post_actuel_int=post
          post_actuel=posterior(a,int)
        }
      }}
    
    
    Mcmc=rbind(Mcmc,c(a,sum(int$unsampled[-(1:(N+2))]),post_actuel))
    
    n=n+1
    if(n==verbose){
      n=0
      print(paste(i,"alpha",Mcmc[i+j,1],"beta",beta,"post",Mcmc[i+j,3]))
    }
  }
  colnames(Mcmc)=c("alpha","nUnsampled","log_post")
  return(list(mcmc=mcmc(Mcmc),post_actuel=post_actuel,a=a,int=int,tree=tree,V=V))
}

